proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1200 @@
+```
+
+The only difference in exports/imports currently is that `push-buffer` and
+`pull-buffer` are represented with a singular `i32` instead of three `i32`
+values. This is because if an export is receiving a buffer it's coming from
+somewhere else which means it's a handle to a buffer and various "intrinsics",
+listed below, will be used to operate on the handle.
+
+## Lifting
+
+The `lift` function here takes four parameters:
+
+* `direction` - one of `"export"` or `"import"` which indicates for which
+  purposes the lifting is happening: lifting an export means lifting
+  a result, lifting an import means lifting a parameter.
+* `src` - this is an instance where the value is being lifted from. This could
+  be a WebAssembly instance or the host itself.
+* `ty` - the interface type that we're lifting into
+* `values` - the core wasm values iterator we have from the function signature
+  or results that we're lifting from. Note that due to type-checking and the
+  canonical ABI for function signatures it's guaranteed that `values` has the
+  right types to lift into `ty`.
+
+This defines how an interface types value is decoded from linear memory and
+WebAssembly values. Note that this function is modeled as a generator. This is
+used later during the definition of `fuse` to model how lifting/lowering is
+interleaved.
+
+```
+def lift(direction, src, ty, values):
+  match ty
+    f32 => yield(values.assert_next(wasm_f32)),
+    f64 => yield(values.assert_next(wasm_f64)),
+
+    s8 => {
+      val = values.assert_next(wasm_i32) as s32
+      assert_or_trap(val >= -128 && val <= 127)
+      yield(val as s8)
+    }
+
+    s16 => {
+      val = values.assert_next(wasm_i32) as s32
+      assert_or_trap(val >= -32768 && val <= 32767)
+      yield(val as s16)
+    }
+
+    u8 => {
+      val = values.assert_next(wasm_i32) as u32
+      assert_or_trap(val >= 0 && val <= 255)
+      yield(val as u8)
+    }
+    u16 => {
+      val = values.assert_next(wasm_i32) as u32
+      assert_or_trap(val >= 0 && val <= 65535)
+      yield(val as u16)
+    }
+
+    s32 => yield(values.assert_next(wasm_i32) as s32),
+    u32 => yield(values.assert_next(wasm_i32) as u32),
+    s64 => yield(values.assert_next(wasm_i64) as s64),
+    u64 => yield(values.assert_next(wasm_i64) as u64),
+
+    char => {
+      codept = values.assert_next(wasm_i32) as u32
+      # generate a trap if `codept` is out of bounds
+      assert_or_trap(codept < 0xD800 || (codpt >= 0xE000 && codept <= 0x10FFFF))
+      yield(codept as char)
+    }
+
+    # Note that lists are only deallocated during lowering for exports. For
+    # imports it's assumed the memory is still owned by the caller WebAssembly
+    # module.
+    #
+    # It's expected that `read_utf8_string` will not trap and will instead
+    # follow the `TextDecoder.decode` semantics which is what WTF-16 languages
+    # like JS/Java/C# will want. This means that invalid utf-8 encodings will
+    # produce replacement characters.
```

Comment:
I am rather unhappy with this paragraph, as I think that we need to account for encoding problems for many popular languages incl. JavaScript properly ([see](https://github.com/WebAssembly/design/issues/1419)). Instead, this paragraph seems to justify to simply accept data corruption, while even doing that in the name of affected languages, which I think is not helpful.

Reply:
Noted.  Probably we shouldn't discuss this broader technical design question here, but in a separate issue and also in your upcoming [June 22 session](https://github.com/WebAssembly/meetings/blob/main/main/2021/CG-06-22.md).  For my part, I'll create a new issue with a more comprehensive justification of the current design choice that a `string` is a `list` of `char` and a `char` is a [Unicode Scalar Value](https://unicode.org/glossary/#unicode_scalar_value), from which this paragraph derives.  (Which [error mode](https://encoding.spec.whatwg.org/#error-mode) we use for `TextDecoder.decode` is a separate question -- I could imagine several viable options here.)

Reply:
Sure, just wanted to make sure that we do not yet merge this PR before the group had the opportunity to discuss implications on real world code. Like, from a purist's angle, I am sympathic like anyone else to the desire of settling Unicode's historic design blunder once and for all with the sledgehammer, but on the other hand I do not think we can when our goals are to reliably port existing real world code, to serve all languages about equally well, to fulfil our promise of establishing a polyglot ecosystem, to integrate seamlessly with the Web platform, to maintain the backwards-compatible nature of the Web and so on, so I do not think Wasm is the right place for this battle to be fought, and perhaps we should just stop fighting it and instead solve the real world technical challenges at hand. Looking forward to your comprehensive justification otherwise, of course.

Reply:
@lukewagner I was about to put my thoughts in the [CG discussion issue](https://github.com/WebAssembly/design/issues/1419), and I remembered your mention here of a justification writeup. Is this still to come? I'd prefer to read it before commenting myself.

Reply:
Sorry for the delay; I've been both chasing down some supporting details and also distracted by other things, but I do want to get to it this week.  Don't feel obligated to wait for me, though.

Reply:
Ok, I posted a first issue summarizing just the first question (about surrogates) in #135.